File: leetcode/c++/Permutation_Sequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        vector<int> elems;
        for(int i = 1; i <= n; i++){
            elems.push_back(i);
        }
        
        return perm_helper(elems, k);
    }
private:
    int fac(int i){
        if(i == 1) return 1;
        int result = 1;
        while(i > 1){
            result *= i;
            i--;
        }
        return result;
    }
    
    string perm_helper(vector<int> &A, int k){
        string result;
        
        if(k == 1){
            for(int i = 0; i < A.size(); i++){
                result += (char)(A[i] + '0');
            }
            return result;
        }
        
        int range = fac(A.size()-1);
        int ind = (k-1)/range;
        
        int new_rank = k%range;
        if(new_rank == 0) new_rank += range;
        
        result += (char)(A[ind]+'0');
        A.erase(A.begin()+ind);
        
        result += perm_helper(A, new_rank);
        return result;
    }
};

File: leetcode/java/PermutationSequence.java
class Solution {
    public String getPermutation(int n, int k) {
        int[] f = factorial(n);
        List<Integer> nums = nums(n);
        String result = "";

        while (n > 1 && k > 1) {
            if (k == f[n - 1] * n) {
                for (int i = nums.size() - 1; i >= 0; i--) {
                    result += nums.get(i);
                }
                return result;
            }

            int index = (k - 1) / f[n - 1];
            Integer x = nums.get(index);
            nums.remove(x);
            result += x;
            if (k != f[n - 1]) k = k - index*f[n-1];
            n--;
        }
        for (Integer x : nums) {
            result += x;
        }
        return result;
    }
    
    private int[] factorial(int n) {
        int[] x = new int[n];
        int f = 1;
        x[0] = 1;
        for(int i = 1; i < n; i++) {
            x[i] = x[i-1]*i;
        }
        return x;
    }
    
    private List<Integer> nums(int n){
        List<Integer> list = new ArrayList<>(n);
        for(int i = 1; i <=n; i++) {
            list.add(i);
        }
        return list;
    }
}
